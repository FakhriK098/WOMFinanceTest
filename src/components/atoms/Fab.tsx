import React from 'react';
import { Image, Pressable, StyleSheet } from 'react-native';
import { colors } from '../../theme/colors';
import { spacing } from '../../theme/spacing';
import { radius } from '../../theme/radius';
import { FABProps } from '../../types/components';

const Fab = ({
  onPress,
  icon,
  style,
  backgroundColor = colors.primary,
  tintColor = colors.white,
}: FABProps) => {
  return (
    <Pressable
      accessibilityRole="button"
      onPress={onPress}
      style={[styles.base, { backgroundColor }, style]}
    >
      <Image
        source={icon}
        style={[styles.icon, { tintColor }]}
        resizeMode="contain"
      />
    </Pressable>
  );
};

const styles = StyleSheet.create({
  base: {
    width: 56,
    height: 56,
    borderRadius: radius.lg,
    alignItems: 'center',
    justifyContent: 'center',
    elevation: 4,
    shadowColor: '#000',
    shadowOpacity: 0.2,
    shadowRadius: 6,
    shadowOffset: { width: 0, height: 3 },
    padding: spacing.sm,
  },
  icon: {
    width: 24,
    height: 24,
  },
});

export default React.memo(Fab);
