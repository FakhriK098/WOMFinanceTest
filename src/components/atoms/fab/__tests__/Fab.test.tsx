import React from 'react';
import renderer, { act } from 'react-test-renderer';

import Fab from '../Fab';

describe('Fab atom', () => {
  it('matches snapshot', () => {
    let tr: renderer.ReactTestRenderer;
    act(() => {
      tr = renderer.create(
        <Fab onPress={() => {}} icon={{ uri: 'icon' }} />,
      );
    });
    expect(tr!.toJSON()).toMatchSnapshot();
    act(() => tr!.unmount());
  });

  it('calls onPress when pressed', () => {
    const onPress = jest.fn();
    let tr: renderer.ReactTestRenderer;
    act(() => {
      tr = renderer.create(
        <Fab onPress={onPress} icon={{ uri: 'icon' }} />,
      );
    });
    const btn = tr!.root.find(
      (inst) => inst.props && inst.props.accessibilityRole === 'button',
    );
    act(() => {
      btn.props.onPress();
    });
    expect(onPress).toHaveBeenCalled();
    act(() => tr!.unmount());
  });

  it('applies background and tint colors', () => {
    let tr: renderer.ReactTestRenderer;
    act(() => {
      tr = renderer.create(
        <Fab
          onPress={() => {}}
          icon={{ uri: 'icon' }}
          backgroundColor="#abcdef"
          tintColor="#112233"
        />,
      );
    });
    const btn = tr!.root.find(
      (inst) => inst.props && inst.props.accessibilityRole === 'button',
    );
    const style = btn.props.style as any[];
    const flattened = Array.isArray(style) ? style : [style];
    expect(flattened.some((s) => s && s.backgroundColor === '#abcdef')).toBe(
      true,
    );
    const image = tr!.root.find((inst) => inst.props && inst.props.resizeMode);
    const imgStyle = image.props.style as any[];
    const imgFlattened = Array.isArray(imgStyle) ? imgStyle : [imgStyle];
    expect(
      imgFlattened.some((s) => s && s.tintColor === '#112233'),
    ).toBe(true);
    act(() => tr!.unmount());
  });
});
