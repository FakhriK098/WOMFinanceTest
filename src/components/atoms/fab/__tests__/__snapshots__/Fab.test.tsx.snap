// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fab atom matches snapshot 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 16,
        "elevation": 4,
        "height": 56,
        "justifyContent": "center",
        "padding": 8,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 6,
        "width": 56,
      },
      {
        "backgroundColor": "#4F46E5",
      },
      undefined,
    ]
  }
>
  <Image
    resizeMode="contain"
    source={
      {
        "uri": "icon",
      }
    }
    style={
      [
        {
          "height": 24,
          "width": 24,
        },
        {
          "tintColor": "#FFFFFF",
        },
      ]
    }
  />
</View>
`;
